### PR TITLE
refactor inference part of videocomposer

### PR DESCRIPTION
**Pull request by zhtmike**
1. remove unnecessary inference (e.g., black image, video encoding)
2. move classifier-free guidance inference into single forward step
3. code refactor
4. unet + scheduler in the graph mode
5. support 910B (online + offline)

### CODE REVIEW

File: examples/videocomposer/vc/models/autoencoder.py [hunk 1]
```diff
@@ -54,6 +54,14 @@ def prune_weights(sd):
                 f"{param_not_load} in network is not loaded or {ckpt_not_load} in checkpoint is not loaded!"
             )
 
+    @ms.jit
```
Comment (SamitHuang) [inline examples/videocomposer/vc/models/autoencoder.py:57]: need to verify it on 910b

Comment (zhtmike) [reply examples/videocomposer/vc/models/autoencoder.py]: reverted the change

File: examples/videocomposer/vc/models/clip_embedder.py
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/models/clip_embedder.py:1]: checked it has no problen

File: examples/videocomposer/vc/engine/inference_single_lite.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            partial_keys = cfg.guidances
+            model_kwargs = prepare_lite_model_kwargs(
+                partial_keys=partial_keys,
+                full_model_kwargs=full_model_kwargs,
+                use_fps_condition=cfg.use_fps_condition,
+            )
+            diffusion_output = diffusion(
+                noise,
+                model_kwargs=model_kwargs,
+                guide_scale=9.0,
```
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/inference_single_lite.py:204]: same here

File: examples/videocomposer/vc/engine/inference_single_lite.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+            end = time.time()
+            time_cost[step, trial] = end - start
+
+            visualize_with_model_kwargs(
+                model_kwargs=model_kwargs,
+                video_data=video_output,
+                ori_video=misc_backups,
+                caps=caps,
+                fname=f"batch_{step}_trial_{trial}.gif",
```
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/inference_single_lite.py:218]: Need to organize the output samples in a better way.
e.g.  `outputs/{task_name}/{video_index}.gif`
```
outputs\
   exp02_motion_transfer\
       vid0001.gif
       vid0001\
            frames\  # if save_frame=True
            inputs.json # {"prompt": xxx, "scale": xxx, "seed": xxx}



```

File: examples/videocomposer/vc/engine/inference_single_lite.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                cfg=cfg,
+            )
+
+    _logger.info("Congratulations! The inference is completed!")
```
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/inference_single_lite.py:222]: add log for where the outputs are saved for easy lookup.

File: examples/videocomposer/vc/engine/modules/decoder.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import mindspore.nn as nn
+import mindspore.ops as ops
+from mindspore import Tensor
+
+
+class Decoder(nn.Cell):
```
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/modules/decoder.py:6]: add short comments like "decode latent frames to pixel frames"? decoder can be a too general name.

File: examples/videocomposer/vc/engine/modules/extractor.py
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/modules/extractor.py:1]: Good modulation

File: examples/videocomposer/vc/engine/utils/helper.py
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/utils/helper.py:1]: i think these are the basic components for the inference pipeline, not just helpers... 
The file structure is a little hard to follow now.
what about improve like this

```
vc\
   model\
      unet_sd.py
   data\
   diffusion\ # only for training
      (gaussian_diffusion.py)  # this file is not used after the inference code refractor.
      latent_diffusion.py
   infer_engine\
       inference_single.py
       inference_multi.py
       schedulers\
           ddim.py
           plms.py
       modules\
           builders.py   # replace helpers.py
           clip.py
           decoder.py
           extractor.py
   trainer\
deploy\
   export.py
   infer_lite.py
   export_modules\
      export_single.py
      export_module.py
  lite_modules\
      ...
  
```

File: examples/videocomposer/vc/engine/utils/visual.py
Comment (SamitHuang) [inline examples/videocomposer/vc/engine/utils/visual.py:1]: better keep it in `vc/utils/`

**Review by SamitHuang — COMMENTED** [review]
code file structures are to be improved.

File: examples/videocomposer/README.md [hunk 6]
```diff
@@ -79,6 +81,72 @@ You can adjust the arguemnts in `vc/config/base.py` (lower-priority) or `configs
...
+
+   `LITE_HOME` is the folder path extracted from tar.gz, and it is recommended to use an absolute path.
+
+   ```shell
+   export LITE_HOME=/path/to/mindspore-lite-{version}-{os}-{platform}
+   export LD_LIBRARY_PATH=$LITE_HOME/runtime/lib:$LITE_HOME/tools/converter/lib:$LD_LIBRARY_PATH
+   export PATH=$LITE_HOME/tools/converter/converter:$LITE_HOME/tools/benchmark:$PATH
+
+
+#### Export Mindspore Lite Model
```
Comment (SamitHuang) [inline examples/videocomposer/README.md:110]: need to unset some environment variables before lite export? Please make sure all necessary instructions for a successful run are included in readme

File: examples/videocomposer/vc/annotator/depth/midas.py [hunk 7]
```diff
@@ -432,7 +437,9 @@ def construct(self, x):
         x = self.head[0](x1)
         size = x.shape[2:]
         size = [s * 2 for s in size]
-        x = ops.interpolate(x, size=size, mode="bilinear", align_corners=True)
+        ori_dtype = x.dtype
+        x = ops.interpolate(x.to(ms.float32), size=size, mode="bilinear", align_corners=True)
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/annotator/depth/midas.py:441]: The change is from #151, will cause error on 910B if we do not modify this

File: examples/videocomposer/README.md [hunk 8]
```diff
@@ -79,6 +81,72 @@ You can adjust the arguemnts in `vc/config/base.py` (lower-priority) or `configs
...
+
+For different task, you can use the corresponding snippet of the code in `run_net.sh`, with extra `--export` flag to save the MindIR model. Please remember to run `export MS_ENABLE_GE=1` first on 910B before running the code snippet.
+
+```shell
+python run_net.py\
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/README.md:115]: sure, will change it

File: examples/videocomposer/vc/engine/export_single.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+def _export_single(cfg: Config) -> None:
+    init(cfg)
+    transforms_list = prepare_transforms(cfg)
+    misc_transforms = transforms_list[1]
+    dataloader = prepare_dataloader(cfg, transforms_list)
+    clip_encoder, clip_encoder_visual = prepare_clip_encoders(cfg)
+    if "y" not in cfg.guidances:
+        zero_y_text = clip_encoder.preprocess([""])
+        zero_y = clip_encoder(zero_y_text)
+        model_export(clip_encoder, [zero_y_text], "clip_encoder_zero_y")
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/export_single.py:56]: it will cause `dynamic shape` problem currently.....

File: examples/videocomposer/vc/config/parser.py [hunk 10]
```diff
@@ -83,6 +77,28 @@ def _parse_args(self):
...
+            "--export",
+            action="store_true",
+            help="save mindir model",
+        )
+        parser.add_argument(
+            "--export_lite",
+            action="store_true",
+            help="save mindspore lite model",
+        )
+        parser.add_argument("--use_lite", action="store_true", help="Use Mindsproe Lite for inference")
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/config/parser.py:101]: yes will move this part

File: examples/videocomposer/vc/engine/modules/sampler.py
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/modules/sampler.py:1]: Will add in another PR

File: examples/videocomposer/vc/engine/utils/helper.py
Comment (zhtmike) [reply examples/videocomposer/vc/engine/utils/helper.py:1]: good suggestion

File: examples/videocomposer/README.md (hunk 6 above)
Comment (zhtmike) [reply examples/videocomposer/README.md:110]: mendtioned in line 128

File: examples/videocomposer/vc/engine/export_single.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+    if cfg.export_lite:
+        gc.collect()
+        lite_converter = LiteConverter(target="ascend")
+        if "y" not in cfg.guidances:
+            lite_converter("clip_encoder_zero_y")
+        if cfg.use_fps_condition:
+            lite_converter("clip_encoder")
+        if not cfg.use_fps_condition:
+            lite_converter("clip_encoder_guidance")
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/export_single.py:222]: same as above

File: examples/videocomposer/vc/annotator/motion/__init__.py [hunk 12]
```diff
@@ -43,7 +43,13 @@ def extract_motion_vectors(input_video, fps=4, viz=False, dump=False, verbose=Fa
     )
     if os.path.exists(tmp_video):
         os.remove(tmp_video)
-    subprocess.run(args=ffmpeg_cmd, shell=True, timeout=120)
+
+    try:
+        subprocess.run(args=ffmpeg_cmd, shell=True, timeout=120, check=True)
+    except Exception as e:
+        raise RuntimeError(
+            "Cannot run ffmpeg, please make sure `ffmpeg` is correctly installed and added to your path."
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/annotator/motion/__init__.py:51]: the `raise xxx from e` where the `from e` will print the original error info

File: examples/videocomposer/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/README.md]: decoupled.

File: examples/videocomposer/vc/engine/export_single.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/export_single.py]: change to inference twice instead.

File: examples/videocomposer/vc/engine/export_single.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+                full_model_kwargs=full_model_kwargs,
+                use_fps_condition=cfg.use_fps_condition,
+            )
+
+            task_model_name = f"{'-'.join(partial_keys)}_sampler"
+
+            diffusion_output = diffusion(
+                noise,
+                model_kwargs=model_kwargs,
+                guide_scale=9.0,
```
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/export_single.py]: done.

File: examples/videocomposer/vc/engine/export_single.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/export_single.py]: done.

File: examples/videocomposer/vc/engine/export_multi.py
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/export_multi.py:1]: deleted.

File: examples/videocomposer/vc/config/parser.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/config/parser.py]: refactored according to this style

File: examples/videocomposer/vc/engine/inference_single_lite.py (hunk 2 above)
Comment (zhtmike) [reply examples/videocomposer/vc/engine/inference_single_lite.py]: done.

File: examples/videocomposer/vc/engine/inference_single_lite.py (hunk 3 above)
Comment (zhtmike) [reply examples/videocomposer/vc/engine/inference_single_lite.py]: organized.

File: examples/videocomposer/vc/engine/inference_single_lite.py (hunk 4 above)
Comment (zhtmike) [reply examples/videocomposer/vc/engine/inference_single_lite.py]: done.

File: examples/videocomposer/vc/engine/modules/decoder.py (hunk 5 above)
Comment (zhtmike) [reply examples/videocomposer/vc/engine/modules/decoder.py]: done

File: examples/videocomposer/vc/engine/utils/functions.py
[the remark this answers is not in the document]
Comment (zhtmike) [reply examples/videocomposer/vc/engine/utils/functions.py:1]: changed to `operators.py`

File: examples/videocomposer/vc/engine/utils/helper.py
Comment (zhtmike) [reply examples/videocomposer/vc/engine/utils/helper.py:1]: Refactored according to this

**Comment by SamitHuang** [comment]
Is the latest commit runnable on 910B + ms2.2? All fp32 is also fine currently

**Comment by zhtmike** [comment]
> Is the latest commit runnable on 910B + ms2.2? All fp32 is also fine currently

Yes. Fine with fp16 pynative and fp32 graph

**Review by SamitHuang — APPROVED** [review]

**Review by SamitHuang — APPROVED** [review]

**Review by SamitHuang — APPROVED** [review]
